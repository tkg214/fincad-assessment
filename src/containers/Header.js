import React, { Component } from 'react';
import { connect } from 'react-redux';

import { login, logout } from '../actions/auth';
import Modal from '../components/Modal';
import Input from '../components/Input';

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = this.defaultClosedModalState;
    this.handleLogoutSubmit = this.handleLogoutSubmit.bind(this);
    this.handleLoginSubmit = this.handleLoginSubmit.bind(this);
    this.handleModalClose = this.handleModalClose.bind(this);
    this.handleModalOpen = this.handleModalOpen.bind(this);
    this.handleLoginInputChange = this.handleLoginInputChange.bind(this);
  }

  get defaultClosedModalState() {
    return {
      showModal: false,
      usernameInput: '',
      passwordInput: ''
    };
  }

  get defaultOpenedModalState() {
    return {
      showModal: true,
      usernameInput: '',
      passwordInput: ''
    };
  }

  handleModalClose() {
    this.setState(this.defaultClosedModalState);
  }

  handleModalOpen() {
    this.setState(this.defaultOpenedModalState);
  }

  handleLogoutSubmit() {
    this.props.logout();
  }

  handleLoginInputChange(e) {
    this.setState({
      [e.target.id]: e.target.value
    });
  }

  handleLoginSubmit(e) {
    e.preventDefault();
    this.props.login(this.state.usernameInput);
    this.setState(this.defaultClosedModalState);
  }

  authed() {
    return (
      <ul className="nav navbar-nav pull-right">
        <li className="nav-item">
          <a data-test="userInfo">
            {`Logged in as ${this.props.auth.name} (${this.props.auth.email})`}
          </a>
        </li>
        <li className="nav-item" onClick={this.handleLogoutSubmit}>
          <a data-test="authed" className="nav-link">
            Logout
          </a>
        </li>
      </ul>
    );
  }

  unauthed() {
    return (
      <ul data-test="unauthed" className="nav navbar-nav pull-right">
        {this.props.auth.userError && this.error()}
        <li className="nav-item" onClick={this.handleModalOpen}>
          <a data-test="unauthed" className="nav-link">
            Login
          </a>
        </li>
      </ul>
    );
  }

  error() {
    return (
      <li className="nav-item">
        <a data-test="error">
          {this.props.auth.errorMessage}
        </a>
      </li>
    );
  }

  render() {
    const { isLoggedIn } = this.props.auth;

    return (
      <div>
        <nav className="navbar navbar-default">
          <div className="container">
            <div className="navbar-brand">FINCAD Tech Assessment</div>
            {isLoggedIn ? this.authed() : this.unauthed()}
          </div>
        </nav>
        <Modal
          closeModal={this.handleModalClose}
          showModal={this.state.showModal}
          modalTitle={'Login'}
        >
          <form onSubmit={this.handleLoginSubmit}>
            <Input
              inputLabel={'Username'}
              inputType={'input'}
              inputId={'usernameInput'}
              inputValue={this.state.usernameInput}
              inputPlaceholder={'Enter Username'}
              handleInputChange={this.handleLoginInputChange}
              data-test="username"
            />
            <Input
              inputLabel={'Password'}
              inputType={'password'}
              inputId={'passwordInput'}
              inputValue={this.state.passwordInput}
              inputPlaceholder={'Enter Password'}
              handleInputChange={this.handleLoginInputChange}
              data-test="password"
            />
            <button
              style={{ marginLeft: '10px' }}
              className="btn pull-right"
              onClick={this.handleModalClose}
            >
              Close
            </button>
            <button className="btn pull-right" type="submit" data-test="submit">
              Login
            </button>
          </form>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  login: username => {
    dispatch(login(username));
  },
  logout: () => {
    dispatch(logout());
  }
});

export { Header as Component };
export default connect(mapStateToProps, mapDispatchToProps)(Header);
