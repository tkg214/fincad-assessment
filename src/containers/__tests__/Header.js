import React from 'react';
import { shallow } from 'enzyme';

import { Component as Header } from '../Header';

test('the header should render correctly', () => {
  const wrapper = shallowRenderHeader();
  expect(wrapper).toMatchSnapshot();
});

test('the logout button and user info should not be displayed if unauthenticated', () => {
  const wrapper = shallowRenderHeader();
  expect(wrapper.find(sel('authed')).node).toBeUndefined;
});

test('the logout button and user should info be displayed if authenticated', () => {
  const auth = {
    isLoggedIn: true,
    name: 'Test Name',
    email: 'test@test.com'
  };
  const wrapper = shallowRenderHeader({ auth });
  expect(wrapper.find(sel('authed')).node).toBeDefined;
  expect(wrapper.find(sel('userInfo')).text()).toBe(
    `Logged in as ${auth.name} (${auth.email})`
  );
});

test('the header displays an error message if there is an error prop', () => {
  const auth = {
    isLoggedIn: false,
    userError: true,
    errorMessage: 'Test Error Message'
  };
  const wrapper = shallowRenderHeader({ auth });
  expect(wrapper.find(sel('error')).text()).toBe(auth.errorMessage);
});

function getHeaderProps(props) {
  return {
    auth: {
      isLoggedIn: false,
      name: '',
      email: '',
      userError: false,
      errorMessage: ''
    },
    ...props
  };
}

function shallowRenderHeader(props = {}) {
  const headerProps = getHeaderProps(props);
  return shallow(<Header {...headerProps} />);
}

function sel(id) {
  return `[data-test="${id}"]`;
}
