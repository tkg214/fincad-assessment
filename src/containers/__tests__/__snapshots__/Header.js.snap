// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the header should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <nav
        className="navbar navbar-default"
    >
        <div
            className="container"
        >
            <div
                className="navbar-brand"
            >
                FINCAD Tech Assessment
            </div>
            <ul
                className="nav navbar-nav pull-right"
                data-test="unauthed"
            >
                <li
                    className="nav-item"
                    onClick={[Function]}
                >
                    <a
                        className="nav-link"
                        data-test="unauthed"
                    >
                        Login
                    </a>
                </li>
            </ul>
        </div>
    </nav>
    <Modal
        closeModal={[Function]}
        modalTitle="Login"
        showModal={false}
    >
        <form
            onSubmit={[Function]}
        >
            <Input
                data-test="username"
                handleInputChange={[Function]}
                inputId="usernameInput"
                inputLabel="Username"
                inputPlaceholder="Enter Username"
                inputType="input"
                inputValue=""
            />
            <Input
                data-test="password"
                handleInputChange={[Function]}
                inputId="passwordInput"
                inputLabel="Password"
                inputPlaceholder="Enter Password"
                inputType="password"
                inputValue=""
            />
            <button
                className="btn pull-right"
                onClick={[Function]}
                style={
                    Object {
                        "marginLeft": "10px",
                      }
                }
            >
                Close
            </button>
            <button
                className="btn pull-right"
                data-test="submit"
                type="submit"
            >
                Login
            </button>
        </form>
    </Modal>
</div>,
  "nodes": Array [
    <div>
      <nav
            className="navbar navbar-default"
      >
            <div
                  className="container"
            >
                  <div
                        className="navbar-brand"
                  >
                        FINCAD Tech Assessment
                  </div>
                  <ul
                        className="nav navbar-nav pull-right"
                        data-test="unauthed"
                  >
                        <li
                              className="nav-item"
                              onClick={[Function]}
                        >
                              <a
                                    className="nav-link"
                                    data-test="unauthed"
                              >
                                    Login
                              </a>
                        </li>
                  </ul>
            </div>
      </nav>
      <Modal
            closeModal={[Function]}
            modalTitle="Login"
            showModal={false}
      >
            <form
                  onSubmit={[Function]}
            >
                  <Input
                        data-test="username"
                        handleInputChange={[Function]}
                        inputId="usernameInput"
                        inputLabel="Username"
                        inputPlaceholder="Enter Username"
                        inputType="input"
                        inputValue=""
                  />
                  <Input
                        data-test="password"
                        handleInputChange={[Function]}
                        inputId="passwordInput"
                        inputLabel="Password"
                        inputPlaceholder="Enter Password"
                        inputType="password"
                        inputValue=""
                  />
                  <button
                        className="btn pull-right"
                        onClick={[Function]}
                        style={
                              Object {
                                    "marginLeft": "10px",
                                  }
                        }
                  >
                        Close
                  </button>
                  <button
                        className="btn pull-right"
                        data-test="submit"
                        type="submit"
                  >
                        Login
                  </button>
            </form>
      </Modal>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Header
        auth={
                Object {
                        "email": "",
                        "errorMessage": "",
                        "isLoggedIn": false,
                        "name": "",
                        "userError": false,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Header {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleLoginInputChange": [Function],
        "handleLoginSubmit": [Function],
        "handleLogoutSubmit": [Function],
        "handleModalClose": [Function],
        "handleModalOpen": [Function],
        "props": Object {
          "auth": Object {
            "email": "",
            "errorMessage": "",
            "isLoggedIn": false,
            "name": "",
            "userError": false,
          },
        },
        "refs": Object {},
        "state": Object {
          "passwordInput": "",
          "showModal": false,
          "usernameInput": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <nav
                    className="navbar navbar-default"
          >
                    <div
                              className="container"
                    >
                              <div
                                        className="navbar-brand"
                              >
                                        FINCAD Tech Assessment
                              </div>
                              <ul
                                        className="nav navbar-nav pull-right"
                                        data-test="unauthed"
                              >
                                        <li
                                                  className="nav-item"
                                                  onClick={[Function]}
                                        >
                                                  <a
                                                            className="nav-link"
                                                            data-test="unauthed"
                                                  >
                                                            Login
                                                  </a>
                                        </li>
                              </ul>
                    </div>
          </nav>
          <Modal
                    closeModal={[Function]}
                    modalTitle="Login"
                    showModal={false}
          >
                    <form
                              onSubmit={[Function]}
                    >
                              <Input
                                        data-test="username"
                                        handleInputChange={[Function]}
                                        inputId="usernameInput"
                                        inputLabel="Username"
                                        inputPlaceholder="Enter Username"
                                        inputType="input"
                                        inputValue=""
                              />
                              <Input
                                        data-test="password"
                                        handleInputChange={[Function]}
                                        inputId="passwordInput"
                                        inputLabel="Password"
                                        inputPlaceholder="Enter Password"
                                        inputType="password"
                                        inputValue=""
                              />
                              <button
                                        className="btn pull-right"
                                        onClick={[Function]}
                                        style={
                                                  Object {
                                                            "marginLeft": "10px",
                                                          }
                                        }
                              >
                                        Close
                              </button>
                              <button
                                        className="btn pull-right"
                                        data-test="submit"
                                        type="submit"
                              >
                                        Login
                              </button>
                    </form>
          </Modal>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <nav
                    className="navbar navbar-default"
          >
                    <div
                              className="container"
                    >
                              <div
                                        className="navbar-brand"
                              >
                                        FINCAD Tech Assessment
                              </div>
                              <ul
                                        className="nav navbar-nav pull-right"
                                        data-test="unauthed"
                              >
                                        <li
                                                  className="nav-item"
                                                  onClick={[Function]}
                                        >
                                                  <a
                                                            className="nav-link"
                                                            data-test="unauthed"
                                                  >
                                                            Login
                                                  </a>
                                        </li>
                              </ul>
                    </div>
          </nav>
          <Modal
                    closeModal={[Function]}
                    modalTitle="Login"
                    showModal={false}
          >
                    <form
                              onSubmit={[Function]}
                    >
                              <Input
                                        data-test="username"
                                        handleInputChange={[Function]}
                                        inputId="usernameInput"
                                        inputLabel="Username"
                                        inputPlaceholder="Enter Username"
                                        inputType="input"
                                        inputValue=""
                              />
                              <Input
                                        data-test="password"
                                        handleInputChange={[Function]}
                                        inputId="passwordInput"
                                        inputLabel="Password"
                                        inputPlaceholder="Enter Password"
                                        inputType="password"
                                        inputValue=""
                              />
                              <button
                                        className="btn pull-right"
                                        onClick={[Function]}
                                        style={
                                                  Object {
                                                            "marginLeft": "10px",
                                                          }
                                        }
                              >
                                        Close
                              </button>
                              <button
                                        className="btn pull-right"
                                        data-test="submit"
                                        type="submit"
                              >
                                        Login
                              </button>
                    </form>
          </Modal>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Header
    auth={
        Object {
            "email": "",
            "errorMessage": "",
            "isLoggedIn": false,
            "name": "",
            "userError": false,
          }
    }
/>,
}
`;
